android/player/src/main/java/com/intuit/playerui/android/asset/DecodableAsset.kt

```diff
@@ -25,4 +48,172 @@ public abstract class DecodableAsset<Data>(assetContext: AssetContext, private v
+    }
+
+    final override fun View.hydrate() {
+        if (this is AsyncViewStub) return
+
+        player.asyncHydrationTrackerPlugin?.trackHydration(this@DecodableAsset)
+        setTag(R.bool.view_hydrated, false)
+        hydrationScope.launch(Dispatchers.Main) { doHydrate() }
+    }
+
+    private suspend fun View.doHydrate() = withContext(Dispatchers.Main) {
+        try {
+            hydrate(getData())
+            setTag(R.bool.view_hydrated, true)
+        } catch (exception: StaleViewException) {
+            // b/c we're launched in a scope that isn't cared about anymore, we can't appropriately handle this, so just fast fail
+            player.inProgressState?.fail(PlayerException("DecodableAssets can't appropriately handle invalidateViews currently, this should be handled in a future major", exception))
+        } finally {
+            player.asyncHydrationTrackerPlugin?.hydrationDone(this@DecodableAsset)
+        }
+    }
+
+    @ExperimentalPlayerApi
+    public class AsyncHydrationTrackerPlugin : AndroidPlayerPlugin {
+        private var trackedHydrations = mutableSetOf<String>()
+
+        public val hooks: Hooks = Hooks()
+
+        public fun trackHydration(asset: DecodableAsset<*>) {
+            synchronized(trackedHydrations) {
+                if (trackedHydrations.isEmpty()) hooks.onHydrationStarted.call()
+                trackedHydrations.add(asset.assetContext.id)
+            }
+        }
+
+        public fun hydrationDone(asset: DecodableAsset<*>) {
+            val doneHydrating = synchronized(trackedHydrations) {
+                trackedHydrations.remove(asset.assetContext.id)
+                trackedHydrations.isEmpty()
+            }
+
+            if (doneHydrating) {
+                hooks.onHydrationComplete.call()
+            }
+        }
+
+        override fun apply(androidPlayer: AndroidPlayer) {
+            androidPlayer.onUpdate { _, _ ->
+                synchronized(trackedHydrations) {
+                    trackedHydrations.clear()
+                }
+            }
+        }
+
+        public class Hooks {
+            public class OnHydrationStartedHook : SyncHook<(HookContext) -> Unit>() {
+                public fun call(): Unit = super.call { f, context ->
+                    f(context)
+                }
+            }
+
+            public class OnHydrationCompleteHook : SyncHook<(HookContext) -> Unit>() {
+                public fun call(): Unit = super.call { f, context ->
+                    f(context)
+                }
+            }
+
+            public val onHydrationStarted: OnHydrationStartedHook = OnHydrationStartedHook()
+            public val onHydrationComplete: OnHydrationCompleteHook = OnHydrationCompleteHook()
+        }
+    }
+
+    /** ViewStub derivative that will replace itself in the view tree once the [view] has resolved */
+    @SuppressLint("ViewConstructor")
+    @InternalPlayerApi
+    public class AsyncViewStub @JvmOverloads constructor(
```

Comment:
If `render` is `suspend`-able, we shouldn't need the async view stub anymore

Reply:
I don't think i'm following this one